prototypes/coin-selection/CoinSelection.hs

```diff
@@ -0,0 +1,278 @@
+    -- ^ Use the /Random-Improve-MA/ algorithm for selecting inputs. This will
+    -- select approximately double the amount necessary for the payment.
+    deriving (Show, Read, Eq, Enum, Generic)
+
+-- | How to assign excess value to change outputs.
+data Change
+    = Balanced
+    -- ^ Value is distributed evenly between change outputs.
+    | Proportional
+    -- ^ Value is distributed between change outputs in proportion to payment
+    -- outputs. Use this with v'RandomImprove' for /emergent/ management of the
+    -- wallet's UTxO.
+    | Unbalanced
+    -- ^ Value is added to the first change output until it reaches the maximum
+    -- size. Then the next change output will receive value.
+    deriving (Show, Read, Eq, Enum, Generic)
+
+-- | Placeholder for the type of possible failure cases in coin selection.
+newtype SelectionError = SelectionError String -- ^ Something went wrong.
+    deriving (Show, Read, Eq, Generic)
+
+-- $naming
+-- We call the coin selection function 'xyzzy' in lieu of a better name.
+--
+-- Possible names could be @balanceTransaction@, @selectCoins@, or
+-- something like that.
+
+-- | The top-level coin selection function.
+--
+-- Given an initial partially funded transaction body, it will produce a
+-- complete, funded, and balanced transaction body from the provided 'UTxO'.
+--
+-- If successful, it returns one or more transactions, otherwise the reason for
+-- failure is provided.
+--
+-- __How to use__
+--
+-- The main interface of 'xyzzy' is its parameter @tx@ the partial
+-- transaction. The caller specifies the outputs of @tx@, and 'xyzzy' calculates
+-- a balanced 'TxBody' from that.
+--
+-- [Payment]: Add a 'TxOut' to the initial @tx@ with the recipient
+--   'Address' and 'Value'.
+-- [Change Output]: Add a 'TxOut' to the initial @tx@ such that the 'Address'
+--   satisfies 'isOurAddress'. Its 'Value' would usually be zero.
+--   If the maximum output size is reached, another 'TxOut' with the same
+--   address will be added to the transaction.
+-- [No new change addresses]: If there is no transaction output that satisfies
+--   v'IsOurAddress', then change will be directed back to the addresses
+--   belonging to the wallet which funded the transaction.
+-- [Join stake pool]: This transaction will involve a delegation certificate,
+--   and possibly also a stake key registration certificate. These cost money.
+--   However 'xyzzy', is not necessarily Cardano-specific or even POS-specific.
+--   Therefore, the 'Address' type for @tx@ should include a constructor for
+--   certificates. The implementation of 'appendTxOutput' should put these
+--   "certificate" address outputs in the correct place within the concrete
+--   'TxBody' type.
+-- [Withdraw from reward account]: Reward account withdrawals add value to the
+--   transaction. The 'TxIn' type for @tx@ should include a constructor for
+--   withdrawals. Put this 'TxIn' either in the 'UTxO' list or in the initial
+--   @tx@.
```

Comment:
> The 'TxIn' type for @tx@ should include a constructor for withdrawals.

👍 

I think the main reason for no making withdrawals equal to `TxIn` on the main chain is that withdrawals could be double-spent, while `TxIn` cannot be double-spent. We may want to consider using a slightly different name, e.g. `FundsIn`. 🤔

Reply:
I think @rvl's idea is to 1) make coin-selection unaware of withdrawals, while 2) still allowing coin-selection to still choose when they should be made.

This is an interesting idea. But:
a. The distinction is important: txs must contain at least one actual `TxIn`
b. Do we really need coin-selection to choose when to make withdrawals?

I think leaving the choice of "should I make a withdrawal or not?" to the caller is fine. This is what we currently [do](https://github.com/input-output-hk/cardano-wallet/blob/ea9f07ce2806d5132b95723136c18c63b63da9f8/lib/core/src/Cardano/Wallet/Transaction.hs#L253). The caller would then add it to the partial `tx`, and coin-selection would not need to be aware of the withdrawal, only its effect on the balance through `evaluateTransactionBalance`.

Reply:
These are good points.